js/bitopro.js

```diff
@@ -0,0 +1,1349 @@
+        const result = [];
+        for (let i = 0; i < markets.length; i++) {
+            const market = markets[i];
+            const active = !this.safeValue (market, 'maintain');
+            const pair = this.safeString (market, 'pair');
+            const baseId = this.safeString (market, 'base');
+            const quoteId = this.safeString (market, 'quote');
+            const id = pair;
+            let base = this.safeCurrencyCode (baseId);
+            base = base.toUpperCase ();
+            let quote = this.safeCurrencyCode (quoteId);
+            quote = quote.toUpperCase ();
+            const symbol = base + '/' + quote;
+            const precision = {
+                'price': this.safeInteger (market, 'quotePrecision'),
+                'amount': this.safeInteger (market, 'basePrecision'),
+            };
+            const limits = {
+                'amount': {
+                    'min': this.safeNumber (market, 'minLimitBaseAmount'),
+                    'max': this.safeNumber (market, 'maxLimitBaseAmount'),
+                },
+                'price': {
+                    'min': undefined,
+                    'max': undefined,
+                },
+                'cost': {
+                    'min': undefined,
+                    'max': undefined,
+                },
+                'leverage': {
+                    'min': undefined,
+                    'max': undefined,
+                },
+            };
+            result.push ({
+                'id': id,
+                'symbol': symbol,
+                'base': base,
+                'quote': quote,
+                'baseId': base,
+                'quoteId': quote,
+                'settle': undefined,
+                'settleId': undefined,
+                'type': 'spot',
+                'spot': true,
+                'margin': false,
+                'swap': false,
+                'future': false,
+                'option': false,
+                'derivative': false,
+                'contract': false,
+                'linear': false,
+                'inverse': false,
+                'contractSize': undefined,
+                'expiry': undefined,
+                'expiryDatetime': undefined,
+                'strike': undefined,
+                'optionType': undefined,
+                'limits': limits,
+                'precision': precision,
+                'active': active,
+                'info': market,
+            });
+        }
+        return result;
+    }
+
+    parseTicker (ticker, market = undefined) {
+        //
+        //     {
+        //         "pair":"btc_twd",
+        //         "lastPrice":"1182449.00000000",
+        //         "isBuyer":false,
+        //         "priceChange24hr":"-1.99",
+        //         "volume24hr":"9.13089740",
+        //         "high24hr":"1226097.00000000",
+        //         "low24hr":"1181000.00000000"
+        //     }
+        //
+        const timestamp = this.milliseconds ();
```

Comment:
we cannot write `timestamp = this.milliseconds()` - it should be undefined if it is not contained in the response